client/src/www/app/outline_server_repository/config.spec.ts

```diff
@@ -0,0 +1,375 @@
+        server: 'example.com',
+        server_port: 443,
+        password: 'PASSWORD',
+      });
+    }).toThrow();
+  });
+
+  it('fails on missing password', () => {
+    expect(() => {
+      config.newTunnelJson({
+        server: 'example.com',
+        server_port: 443,
+        method: 'METHOD',
+      });
+    }).toThrow();
+  });
+
+  it('parses new format', () => {
+    expect(
+      config.newTunnelJson({
+        transport: {
+          type: 'shadowsocks',
+          endpoint: {
+            type: 'dial',
+            host: 'example.com',
+            port: 443,
+          },
+          cipher: 'METHOD',
+          secret: 'PASSWORD',
+          prefix: '\x03\x02\x03',
+        },
+      })
+    ).toEqual({
+      transport: {
+        type: 'shadowsocks',
+        endpoint: {
+          type: 'dial',
+          host: 'example.com',
+          port: 443,
+        },
+        cipher: 'METHOD',
+        secret: 'PASSWORD',
+        prefix: '\x03\x02\x03',
+      },
+    } as config.TunnelConfigJson);
+  });
+
+  it('parses abbreviated endpoint', () => {
+    expect(
+      config.newTunnelJson({
+        transport: {
+          type: 'shadowsocks',
+          endpoint: 'example.com:443',
+          cipher: 'METHOD',
+          secret: 'PASSWORD',
+          prefix: '\x03\x02\x03',
+        },
+      })
+    ).toEqual({
+      transport: {
+        type: 'shadowsocks',
+        endpoint: {
+          type: 'dial',
+          host: 'example.com',
+          port: 443,
+        },
+        cipher: 'METHOD',
+        secret: 'PASSWORD',
+        prefix: '\x03\x02\x03',
+      },
+    } as config.TunnelConfigJson);
+  });
+
+  it('parses multi-hop', () => {
+    expect(
+      config.newTunnelJson({
+        transport: {
+          type: 'shadowsocks',
+          endpoint: {
+            type: 'dial',
+            host: 'server2.com',
+            port: 443,
+            dialer: {
+              type: 'shadowsocks',
+              endpoint: 'server1.com:8888',
+              cipher: 'METHOD1',
+              secret: 'PASSWORD1',
+              prefix: '\x01\x01\x01',
+            },
+          },
+          cipher: 'METHOD2',
+          secret: 'PASSWORD2',
+          prefix: '\x03\x02\x03',
+        },
+      })
+    ).toEqual({
+      transport: {
+        type: 'shadowsocks',
+        endpoint: {
+          type: 'dial',
+          host: 'server2.com',
+          port: 443,
+          dialer: {
+            type: 'shadowsocks',
+            endpoint: {
+              type: 'dial',
+              host: 'server1.com',
+              port: 8888,
+            },
+            cipher: 'METHOD1',
+            secret: 'PASSWORD1',
+            prefix: '\x01\x01\x01',
+          },
+        },
+        cipher: 'METHOD2',
+        secret: 'PASSWORD2',
+        prefix: '\x03\x02\x03',
+      },
+    } as config.TunnelConfigJson);
+  });
+
+  it('parses pipe', () => {
+    expect(
+      config.newTunnelJson({
+        transport: [
```

Comment:
This is the pipe format, similar to the SDK format, but in JSON. It simplified the nesting. The idea is that one entry replaces the previous as the "default" dialer.